File: examples/integration/governance_guard.py
"""
Morrison Runtime Governance — drop-in pre-execution guard (Python).

One synchronous call at your agent's plan -> act boundary. The governance
service NEVER executes your tool; it inspects the proposed tool call and returns
a deterministic verdict before you run anything:

    PERMIT   -> execute the tool
    ESCALATE -> hold for a human (sign-off); a review card says who + why
    BLOCK    -> deny before execution

Config (env):
    GOVERNANCE_URL    base URL of the service (hosted or self-hosted),
                      e.g. https://your-governance-host  (default http://localhost:8000)
    GOVERNANCE_TOKEN  optional Bearer token if the service requires auth

    pip install requests
"""
from __future__ import annotations

import os
from typing import Any, Optional

import requests

GOVERNANCE_URL = os.environ.get("GOVERNANCE_URL", "http://localhost:8000").rstrip("/")
GOVERNANCE_TOKEN = os.environ.get("GOVERNANCE_TOKEN")

ToolCall = dict[str, Any]  # {"tool": str, "args": {...}}


class GovernanceBlocked(Exception):
    """Raised when a trajectory reaches a forbidden state Ω (verdict BLOCK)."""

    def __init__(self, result: dict):
        self.result = result
        super().__init__(result.get("reason", "Blocked before execution"))


class GovernanceEscalation(Exception):
    """Raised when a trajectory needs human sign-off (verdict ESCALATE)."""

    def __init__(self, result: dict):
        self.result = result
        self.review = result.get("review", {})
        super().__init__(self.review.get("reason", "Held for human review"))


def governance_check(trajectory: list[ToolCall], *, domains: Optional[list[str]] = None,
                     horizon: Optional[int] = None, timeout: float = 4.0) -> dict:
    """Evaluate a proposed trajectory. Returns the raw GovernanceResult dict:
    {verdict, permitted, blocked, layer, reason, omega_domain, trajectory_hash,
     attestation, requires_human_review?, review?}."""
    headers = {"content-type": "application/json"}
    if GOVERNANCE_TOKEN:
        headers["authorization"] = f"Bearer {GOVERNANCE_TOKEN}"
    body: dict[str, Any] = {"trajectory": trajectory}
    if domains:
        body["domains"] = domains
    if horizon:
        body["horizon"] = horizon
    resp = requests.post(f"{GOVERNANCE_URL}/v1/evaluate", json=body, headers=headers, timeout=timeout)
    resp.raise_for_status()
    return resp.json()


def guard(tool: str, args: Optional[dict] = None, *, domains: Optional[list[str]] = None) -> dict:
    """Call this immediately BEFORE you execute a tool.

    Returns the verdict dict on PERMIT; raises GovernanceEscalation on ESCALATE
    and GovernanceBlocked on BLOCK — so a denied/held action never runs.
    """
    result = governance_check([{"tool": tool, "args": args or {}}], domains=domains)
    if result.get("permitted"):
        return result
    if result.get("verdict") == "ESCALATE":
        raise GovernanceEscalation(result)
    raise GovernanceBlocked(result)


# ── LangChain: govern any tool so every invocation is checked pre-execution ──
def govern_langchain_tool(tool):  # type: ignore[no-untyped-def]
    """Wrap a LangChain BaseTool so `guard(...)` runs before `tool._run`."""
    original = tool._run

    def _run(*args, **kwargs):
        guard(tool.name, kwargs or {"args": list(args)})  # raises on BLOCK / ESCALATE
        return original(*args, **kwargs)

    tool._run = _run
    return tool


if __name__ == "__main__":
    # Example: a high-value transfer is blocked before execution.
    try:
        guard("transfer_funds", {"amount": 50000, "to": "acct_991"}, domains=["finance"])
        print("ALLOW — safe to execute")
    except GovernanceEscalation as e:
        print("ESCALATE —", e.review.get("decision_authority"), "·", e)
    except GovernanceBlocked as e:
        print("BLOCK —", e)

File: examples/integration/governanceGuard.ts
/**
 * Morrison Runtime Governance — drop-in pre-execution guard (TypeScript).
 *
 * One async call at your agent's plan -> act boundary. The governance service
 * NEVER executes your tool; it inspects the proposed tool call and returns a
 * deterministic verdict before you run anything:
 *
 *   PERMIT   -> execute the tool
 *   ESCALATE -> hold for a human (sign-off); a review card says who + why
 *   BLOCK    -> deny before execution
 *
 * Env:
 *   GOVERNANCE_URL    base URL of the service (hosted or self-hosted)
 *   GOVERNANCE_TOKEN  optional Bearer token if the service requires auth
 */

export type Verdict = "PERMIT" | "ESCALATE" | "BLOCK";

export interface ReviewCard {
  reason: string;
  required_action: string;
  decision_authority: string;
  next_step: string;
  execution_status: string; // "HELD FOR HUMAN REVIEW"
}

export interface GovernanceResult {
  verdict: Verdict;
  permitted: boolean;
  blocked: boolean;
  layer: string;
  reason: string;
  omega_domain: string | null;
  trajectory_hash: string;
  requires_human_review?: boolean;
  review?: ReviewCard;
  attestation?: { engine_commit: string; ruleset_hash: string };
}

export interface ToolCall {
  tool: string;
  args?: Record<string, unknown>;
}

const BASE_URL = (process.env.GOVERNANCE_URL ?? "http://localhost:8000").replace(/\/$/, "");
const TOKEN = process.env.GOVERNANCE_TOKEN;

export class GovernanceBlocked extends Error {
  constructor(public result: GovernanceResult) { super(result.reason); }
}
export class GovernanceEscalation extends Error {
  constructor(public result: GovernanceResult) { super(result.review?.reason ?? "Held for human review"); }
}

/** Evaluate a proposed trajectory (one or more tool calls) before execution. */
export async function governanceCheck(
  trajectory: ToolCall[],
  opts: { domains?: string[]; horizon?: number; timeoutMs?: number } = {},
): Promise<GovernanceResult> {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), opts.timeoutMs ?? 4000);
  try {
    const res = await fetch(`${BASE_URL}/v1/evaluate`, {
      method: "POST",
      headers: { "content-type": "application/json", ...(TOKEN ? { authorization: `Bearer ${TOKEN}` } : {}) },
      body: JSON.stringify({ trajectory, domains: opts.domains, horizon: opts.horizon }),
      signal: ctrl.signal,
    });
    if (!res.ok) throw new Error(`governance service ${res.status}`);
    return (await res.json()) as GovernanceResult;
  } finally {
    clearTimeout(t);
  }
}

/**
 * Gate a single tool dispatch. Call this right before your agent executes a
 * tool. Runs the tool only on PERMIT; routes to `onEscalate` on ESCALATE;
 * throws GovernanceBlocked on BLOCK — so a denied/held action never runs.
 */
export async function guardedDispatch<T>(
  call: ToolCall,
  run: (call: ToolCall) => Promise<T> | T,
  onEscalate?: (r: GovernanceResult) => Promise<T> | T,
  opts: { domains?: string[] } = {},
): Promise<T> {
  const v = await governanceCheck([call], opts);
  if (v.permitted) return run(call);                       // ALLOW
  if (v.verdict === "ESCALATE") {
    if (onEscalate) return onEscalate(v);                  // ESCALATE -> human
    throw new GovernanceEscalation(v);
  }
  throw new GovernanceBlocked(v);                          // BLOCK
}
